core/src/test/scala/integration/kafka/api/OffsetOfMaxTimestampTest.scala

```diff
@@ -0,0 +1,77 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package integration.kafka.api
+
+import kafka.api.IntegrationTestHarness
+import kafka.utils.{TestInfoUtils, TestUtils}
+import org.apache.kafka.clients.CommonClientConfigs
+import org.apache.kafka.clients.admin.{Admin, NewTopic, OffsetSpec}
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.config.TopicConfig
+import org.apache.kafka.common.serialization.StringSerializer
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+
+import java.util.{Collections, Properties}
+
+class OffsetOfMaxTimestampTest extends IntegrationTestHarness {
+  @AfterEach
+  override def tearDown(): Unit = {
+    TestUtils.shutdownServers(brokers)
+    super.tearDown()
+  }
+
+  override def brokerCount: Int = 1
+
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testWithNoCompression(quorum: String): Unit = {
+    test(false)
+  }
+
+  private def test(useCompression: Boolean): Unit = {
+    val topicName: String = "OffsetOfMaxTimestampTest-" + System.currentTimeMillis()
+
+    val admin: Admin = Admin.create(adminClientConfig)
+    try {
+      admin.createTopics(Collections.singletonList(new NewTopic(topicName, 1, 1.toShort)
+        .configs(Collections.singletonMap(TopicConfig.COMPRESSION_TYPE_CONFIG, if (useCompression) "gzip" else "none"))))
+      val props: Properties = new Properties
+      props.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers())
+      props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer])
+      props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer])
+      props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, if (useCompression) "gzip" else "none")
+      val producer: KafkaProducer[String, String] = new KafkaProducer[String, String](props)
+      try {
+        val time: Long = 10000
+        producer.send(new ProducerRecord[String, String](topicName, 0, time + 100, null, "val20"))
+        producer.send(new ProducerRecord[String, String](topicName, 0, time + 400, null, "val15"))
+        producer.send(new ProducerRecord[String, String](topicName, 0, time + 250, null, "val15"))
```

Comment:
How do we make sure that the 3 records are in the same batch?

Reply:
https://github.com/apache/kafka/pull/15474#discussion_r1518055609 can address that.

Reply:
Do we need a separate test for this? I mentioned how to test here: https://github.com/apache/kafka/pull/15474#discussion_r1516625703

Reply:
For the non-compression case, we can just make the record size 10 B (rather than 10KB). 
There is also some discussion about how to control the batch size via flush

Reply:
@jolshan I prefer to have following test: https://github.com/apache/kafka/pull/15474#discussion_r1520206161